runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -1187,24 +1187,29 @@ function M.reset(client_id, buffer_client_map)
   end)
 end
 
---- Sets the location list
----@param opts table|nil Configuration table. Keys:
----         - {open_loclist}: (boolean, default true)
----             - Open loclist after set
+
+--- Gets the diagnostics in a format that is compatible with the
+--- quickfix-/locationlist
+---
+--- To populate the quickfix list, you can use this function like so:
+---
+---     vim.fn.setqflist({}, 'r', {
+---       title = 'Language Server';
+---       items = vim.lsp.diagnostic.get_as_qfitems { workspace = true }
+---     })
+---
+---@param opts table|nil Configuration table.
 ---         - {client_id}: (number)
 ---             - If nil, will consider all clients attached to buffer.
 ---         - {severity}: (DiagnosticSeverity)
 ---             - Exclusive severity to consider. Overrides {severity_limit}
 ---         - {severity_limit}: (DiagnosticSeverity)
 ---             - Limit severity of diagnostics found. E.g. "Warning" means { "Error", "Warning" } will be valid.
 ---         - {workspace}: (boolean, default false)
----             - Set the list with workspace diagnostics
-function M.set_loclist(opts)
+---             - Include workspace diagnostics. If false (default) only
+---               diagnostics for the current buffer are included.
+function M.get_as_qfitems(opts)
```

Comment:
I'm not sure about the name. Happy to change it to something else if there's a good suggestion.

Reply:
Maybe something like `locations_to_items` would be more descriptive? And possibly put it in `util.lua`?

As an aside, I see the LSP API as basically having three layers:
1. the server-facing layer, which in naming, data structures and handling should follow LSP terminology as closely as possible (talking about `Locations` and `Tokens` and whatnot) -- these are functions in `lsp.lua` etc.
2. the vim-facing layer, which should "speak vim" -- `clist`, `loclist` etc. -- these are functions in `buf.lua` and `diagnostic.lua`
3. a set of (exposed) utility functions that translate between the two and live in `util.lua`

Reply:
Yes I was considering a `diagnostic_to_items` in `util.lua` first. It's also what I have in my dotfiles https://github.com/mfussenegger/dotfiles/blob/579ab91da5b14ba20766f2acc11d855a56aa0a5f/vim/.config/nvim/lua/me/lsp/diagnostic.lua#L5-L25 it takes the diagnostic grouped by bufnr and a predicate -  you can use it like this:

```
  local items = diagnostics_to_items(
    vim.lsp.diagnostic.get_all(),
    function(d) return d.severity == vim.lsp.protocol.DiagnosticSeverity.Error end
  )
```

I'd prefer that variant (and trimming the amount of supported options). The reason I went with this `get_as_qfitems` is mostly because of how `set_loclist` already supports a lot of options.

I'll wait and see what the others have to say.

Reply:
How would you feel about merging this as-is. Would `diagnostics_to_items` be an ok name here, and then we could migrate the function to a utility function?

Reply:
Pushed a fixup with the rename. Let me know if it is otherwise good to go - I'll squash before it gets merged.

Reply:
Actually, let me try another variant first.

Reply:
I pushed another fixup - adding a `util.diagnostics_to_items`. It also includes a fix for `set_loclist` (previously if `workspace` was `true`, the `client_id` option was ignored)

Let me know which variant you prefer, the initial variant in the first commit - or the one in the last commit.